lib/pg_ha_migrations/unsafe_statements.rb

```diff
@@ -1,114 +1,85 @@
 module PgHaMigrations::UnsafeStatements
-  def self.delegate_unsafe_method_to_migration_base_class(method_name)
-    define_method("unsafe_#{method_name}") do |*args, &block|
-      # Dispatching here is a bit complicated: we need to execute the method
-      # belonging to the first member of the inheritance chain (besides
-      # UnsafeStatements). If don't find the method in the inheritance chain,
-      # we need to rely on the ActiveRecord::Migration#method_missing
-      # implementation since much of ActiveRecord::Migration's functionality
-      # is not implemented in real methods but rather by proxying.
-      #
-      # For example, ActiveRecord::Migration doesn't define #create_table.
-      # Instead ActiveRecord::Migration#method_missing proxies the method
-      # to the connection. However some migration compatibility version
-      # subclasses _do_ explicitly define #create_table, so we can't rely
-      # on only one way of finding the proper dispatch target.
-
-      # Exclude our `raise` guard implementations.
-      ancestors_without_unsafe_statements = self.class.ancestors - [PgHaMigrations::UnsafeStatements]
-
-      delegate_method = self.method(method_name)
-      candidate_method = delegate_method
-
-      # Find the first usable method in the ancestor chain
-      # or stop looking if there are no more possible
-      # implementations.
-      until candidate_method.nil? || ancestors_without_unsafe_statements.include?(candidate_method.owner)
-        candidate_method = candidate_method.super_method
-      end
-
-      if candidate_method
-        delegate_method = candidate_method
+  def self.disable_or_delegate_default_method(method_name, error_message, delegate_from_compatibility: false)
+    define_method(method_name) do |*args, &block|
+      if PgHaMigrations.config.disable_default_migration_methods
+        unless  delegate_from_compatibility && caller[0] =~ /lib\/active_record\/migration\/compatibility.rb/
```

Comment:
Perhaps instead of `delegate_from_compatibility` this would be better named something like `allow_reentry_from_compatibility_module`?

Also, I realize this was originally my code (and I should have done this earlier), but could we add a comment like:

```
# Most migration methods are only ever called by a migration and
# therefore aren't re-entrant or callable from another migration
# method, but `execute` is called directly by at least one of the
# implementations in `ActiveRecord::Migration::Compatibility` so
# we have to explicitly handle that case by allowing execution of
# the original implementation by its original name.
```